core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/BestSolutionHolder.java

```diff
@@ -0,0 +1,151 @@
+    /**
+     * Sets the new best solution if all known problem changes have been processed and thus are contained in this
+     * best solution.
+     * 
+     * @param bestSolution the new best solution that replaces the previous one if there is any
+     * @param isEveryProblemChangeProcessed a supplier that tells if all problem changes have been processed
+     */
+    void set(Solution_ bestSolution, Supplier<Boolean> isEveryProblemChangeProcessed) {
+        problemChangesLock.lock();
+        try {
+            /*
+             * The new best solution can be accepted only if there are no pending problem changes nor any additional
+             * changes may come during this operation. Otherwise, a race condition might occur that leads to associating
+             * problem changes with a solution that was created later, but does not contain them yet.
+             * As a result, CompletableFutures representing these changes would be completed too early.
+             */
+            if (isEveryProblemChangeProcessed.get()) {
+                versionedBestSolutionRef.set(new VersionedBestSolution(bestSolution, currentVersion));
+                currentVersion = currentVersion.add(BigInteger.ONE);
+            }
+        } finally {
+            problemChangesLock.unlock();
+        }
+    }
+
+    /**
+     * Adds a new problem change to a solver and registers the problem change to be later retrieved together with
+     * a relevant best solution by the {@link #take()} method.
+     * 
+     * @return CompletableFuture that will be completed after the best solution containing this change is passed to
+     *         a user-defined Consumer.
+     */
+    CompletableFuture<Void> addProblemChange(Solver<Solution_> solver, ProblemChange<Solution_> problemChange) {
+        CompletableFuture<Void> futureProblemChange;
+        problemChangesLock.lock();
+        try {
+            futureProblemChange = new CompletableFuture<>();
+            problemChangesPerVersion.compute(currentVersion, (version, futureProblemChangeList) -> {
+                if (futureProblemChangeList == null) {
+                    futureProblemChangeList = new ArrayList<>();
+                }
+                futureProblemChangeList.add(futureProblemChange);
+                return futureProblemChangeList;
+            });
+            solver.addProblemChange(problemChange);
+        } finally {
+            problemChangesLock.unlock();
+        }
+
+        return futureProblemChange;
+    }
+
+    void cancelPendingChanges() {
+        problemChangesLock.lock();
+        try {
+            problemChangesPerVersion.values()
+                    .stream()
+                    .flatMap(Collection::stream)
+                    .forEach(pendingProblemChange -> pendingProblemChange.cancel(false));
+            problemChangesPerVersion.clear();
+        } finally {
+            problemChangesLock.unlock();
+        }
+    }
+
+    private final class VersionedBestSolution {
```

Comment:
I'd make it static. You'll have to reintroduce `<Solution_>`, but that's not an issue, I think.

Reply:
Good point.